dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/RepeatableStageParamsHelper.java

```diff
@@ -0,0 +1,264 @@
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.analytics.util;
+
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import org.hisp.dhis.common.RepeatableStageParams;
+import org.hisp.dhis.common.exception.InvalidRepeatableStageParamsException;
+import org.hisp.dhis.period.Period;
+import org.hisp.dhis.period.RelativePeriodEnum;
+import org.hisp.dhis.period.RelativePeriods;
+import org.hisp.dhis.util.DateUtils;
+
+public class RepeatableStageParamsHelper
+{
+    private static final String PS_INDEX_REGEX = "\\[-?\\d+\\]";
+
+    private static final String PS_ASTERISK_REGEX = "\\[\\*\\]";
+
+    private static final String PS_INDEX_COUNT_REGEX = "\\[-?\\d+,\\s*\\d+\\]";
+
+    private static final String PS_INDEX_COUNT_START_DATE_END_DATE_REGEX = "\\[-?\\d+,\\s*\\d+,\\s*\\d{4}\\-(0[1-9]|1[012])\\-(0[1-9]|[12][0-9]|3[01]),\\s*\\d{4}\\-(0[1-9]|1[012])\\-(0[1-9]|[12][0-9]|3[01])\\]";
+
+    private static final String PS_INDEX_COUNT_RELATIVE_PERIOD_REGEX = "\\[-?\\d+,\\s*\\d+,\\s*\\w+\\]";
+
+    private static final String PS_START_DATE_END_DATE_REGEX = "\\[\\d{4}\\-(0[1-9]|1[012])\\-(0[1-9]|[12][0-9]|3[01]),\\s*\\d{4}\\-(0[1-9]|1[012])\\-(0[1-9]|[12][0-9]|3[01])\\]";
+
+    private static final String PS_RELATIVE_PERIOD_REGEX = "\\[\\w+\\]";
+
+    private static final Pattern[] PS_PARAMS_PATTERN_LIST = {
+        Pattern.compile( PS_INDEX_REGEX ),
+        Pattern.compile( PS_ASTERISK_REGEX ),
+        Pattern.compile( PS_INDEX_COUNT_REGEX ),
+        Pattern.compile( PS_INDEX_COUNT_START_DATE_END_DATE_REGEX ),
+        Pattern.compile( PS_INDEX_COUNT_RELATIVE_PERIOD_REGEX ),
+        Pattern.compile( PS_START_DATE_END_DATE_REGEX ),
+        Pattern.compile( PS_RELATIVE_PERIOD_REGEX )
+    };
+
+    private RepeatableStageParamsHelper()
+    {
+    }
+
+    public static RepeatableStageParams getRepeatableStageParams( String dimension )
```

Comment:
Would be nice to document all the methods in this class, including the private ones. So we know what they are doing and what's the expected return format.